[DX-1278] Gateway release notes 5.3.1 (#4399)
TykTechnologies/tyk-docs: tyk-docs/content/product-stack/tyk-gateway/release-notes/version-5.3.md

---

## 5.3.1 Release Notes

### Release Date TBC

### Breaking Changes
**Attention**: Please read this section carefully.

There are no breaking changes in this release, however if moving from an version of Tyk older than 5.3.0 please read the explanation provided with [5.3.0 release]({{< ref "#TykOAS-v5.3.0">}}).

### Deprecations
There are no deprecations in this release.

### Upgrade Instructions
If you are on a 5.3.0 we advise you to upgrade ASAP and if you are on an older version skip 5.3.0 and upgrade directly to this release. Go to the [Upgrading Tyk](#upgrading-tyk) section for detailed upgrade instructions.

### Release Highlights
This release primarily focuses on bug fixes.
For a comprehensive list of changes, please refer to the detailed [changelog]({{< ref "#Changelog-v5.3.1">}}) below.

### Downloads
- [Docker image to pull](TBC)
- [source code](TBC)

### Changelog {#Changelog-v5.3.1}

#### Fixed

<ul>
<li>
<details>
<summary>Security enhancement: don't load APIs into Gateway if custom plugin bundle fails to load</summary>

> Issues were addressed where Tyk failed to properly reject custom plugin bundles with signature verification failures, allowing APIs to load without necessary plugins, potentially exposing upstream services. With the fix, if plugin bundle loading fails (e.g., due to signature verification), the API will not be loaded and an error will be logged in the Gateway.

Author: ```suggestion
Issues were addressed where Tyk failed to properly reject custom plugin bundles with signature verification failures, allowing APIs to load without necessary plugins, potentially exposing upstream services. With the fix, if the plugin bundle fails to load (for example, due to failed signature verification) the API will not be loaded and an error will be logged in the Gateway.
```

</details>
</li>
<li>
<details>
<summary>Policy-API link deletion code updated to ensure DocumentDB compatibility</summary>

Policy-API link deletion code was updated to address a compatibility issue with DocumentDB. The previous version relied on the $expr operator, supported by MongoDB but not DocumentDB.
</details>
</li>
<li>
<details>
<summary>Fixed a Gateway panic that could occur when using custom JavaScript plugins with the Ignore Authentication middleware</summary>

Fixed a panic scenario that occurred when a custom JavaScript plugin that requests access to the session metadata (`require_session:true`) is assigned to the same endpoint as the Ignore Authentication middleware. While the custom plugin expects access to a valid session, the configuration flag doesn't guarantee its presence, only that it's passed if available. As such, the custom plugin should be coded to verify that the session metadata is present before attempting to use it.
</details>
</li>
<li>
<details>
<summary>Tyk cache confusion leads to mixed GraphQL body responses</summary>

> GraphQL APIs were returning incorrect responses when simultaneous GQL calls with different inputs were made, which looked as if caching was mixed up. This was related to a setting in the GraphQL engine, that has now been turned off, so simultaneous GQL calls won't return incorrect response

Author: ```suggestion
GraphQL APIs were returning incorrect responses when simultaneous GraphQL calls with different inputs were made, which looked as if caching was mixed up. This was related to a setting in the GraphQL engine, that has now been turned off, so simultaneous calls won't return an incorrect response.
```

</details>
</li>
<li>
<details>
<summary>Gateway panics when arguments are missing in persist GraphQL endpoints</summary>

> In some instances users were noticing gateway panics when using "Persist GQL operation middleware" without arguments defined. This issue has been fixed and the gateway will not throw panics in these cases anymore.

Author: ```suggestion
In some instances users were noticing gateway panics when using the "Persist GQL" middleware without arguments defined. This issue has been fixed and the gateway will not throw panics in these cases anymore.
```

</details>
</li>
<li>
<details>

> <summary>Incorrect MDCB Gateway Counter on License Statistics Page</summary>

Reviewer: ```suggestion
<summary>Incorrect MDCB Gateway Counter on License Statistics page</summary>
```

Author: ```suggestion
<summary>Tyk Dashboard License Statistics page could display incorrect number of data plane gateways</summary>
```

> Resolved an issue where the MDCB gateway counter inaccurately incremented when a gateway was stopped and restarted.

Author: ```suggestion
Resolved an issue in distributed deployments where the MDCB data plane gateway counter was inaccurately incremented when a Gateway was stopped and restarted.
```

</details>
</li>
<li>
<details>

> <summary>Missing GraphQL OTel Attributes in Spans on Request Validation Failure</summary>

Reviewer: ```suggestion
<summary>Missing GraphQL OTel attributes in Spans on Request Validation failure</summary>
```

Author: ```suggestion
<summary>Missing GraphQL OTel attributes in spans when requests fail validation</summary>
```

> In cases where detailed_tracing was set to false and client was sending a malformed request to a GQL API the traces were missing GQL attributes (operation name, type and document). This has been corrected and debugging GQL with OTel will be easier for users.

Author: ```suggestion
In cases where `detailed_tracing` was set to `false` and the client was sending a malformed request to a GraphQL API, the traces were missing GraphQL attributes (operation name, type and document). This has been corrected and debugging GraphQL with OTel will be easier for users.
```

</details>
</li>
<li>
<details>

> <summary>API Cache Clearing from Dashboard Not Propagated to Edge Deployments</summary>

Reviewer: ```suggestion
<summary>API cache clearing from Dashboard not propagated to edge deployments</summary>
```

Author: ```suggestion
<summary>Unable to clear the API cache in distributed data plane Gateways from the control plane Dashboard</summary>
```

> Addressed a bug where clearing the API cache from the Tyk Dashboard UI failed to invalidate the cache in distributed data plane gateways.

Author: ```suggestion
Addressed a bug where clearing the API cache from the Tyk Dashboard failed to invalidate the cache in distributed data plane gateways.
```

---

## 5.3.1 Release Notes

### Release Date 24 April 2024

### Breaking Changes
**Attention**: Please read this section carefully.

There are no breaking changes in this release, however if moving from an version of Tyk older than 5.3.0 please read the explanation provided with [5.3.0 release]({{< ref "#TykOAS-v5.3.0">}}).

### Deprecations
There are no deprecations in this release.

### Upgrade Instructions

> If you are using 5.3.0 we advise you to upgrade ASAP and if you are on an older version you should skip 5.3.0 and upgrade directly to this release. Go to the [Upgrading Tyk](#upgrading-tyk) section for detailed upgrade instructions.

Reviewer 2: If user is upgrading from older version and skip 5.3.0, they need to review and follow 5.3.0 upgrade instructions too.

Reviewer: @caroltyk I have updated, if this and other comments are ok could you approve for release?

### Release Highlights
This release primarily focuses on bug fixes.
For a comprehensive list of changes, please refer to the detailed [changelog]({{< ref "#Changelog-v5.3.1">}}) below.

### Dependencies

Please refer to [dependencies](#dependencies-5.3.0) provided for release 5.3.0

### Downloads
- [Docker image to pull](https://hub.docker.com/r/tykio/tyk-gateway/tags?page=&page_size=&ordering=&name=v5.3.1)
  - ```bash
    docker pull tykio/tyk-gateway:v5.3.1
    ``` 
- Helm charts

>   - [tyk-charts GH Repo](https://github.com/TykTechnologies/tyk-charts/releases)

Reviewer 2: Latest helm chart release notes is in Docs: /product-stack/tyk-charts/release-notes/version-1.3

Reviewer: @caroltyk I have updated PR to mention MDCB 2.5.1. If ok and other comments ok would be to approve for release?

</details>
</li>
<li>
<details>
<summary>Incorrect naming for semantic conventions attributes in GQL spans</summary>

GQL Open Telemetry semantic conventions attribute names were missing `graphql` prefix and therefore were not in line with the community standard. This has been fixed and all attributes have the correct prefix.
</details>
</li>
<li>
<details>
<summary>URL Rewrite middleware did not always correctly observe quotas for requests using keys created from policies</summary>

Fixed two bugs in the handling of usage quotas by the URL rewrite middleware when it was configured to rewrite to itself (e.g. to `tyk://self`). Quota limits were not observed and the quota related response headers always contained `0`. 
</details>
</li>
<li>
<details>
<summary>Tyk Dashboard License Statistics page could display incorrect number of data plane gateways</summary>

Resolved an issue in distributed deployments where the MDCB data plane gateway counter was inaccurately incremented when a Gateway was stopped and restarted.
</details>
</li>
<li>
<details>
<summary>Unable to clear the API cache in distributed data plane Gateways from the control plane Dashboard</summary>

> Addressed a bug where clearing the API cache from the Tyk Dashboard failed to invalidate the cache in distributed data plane gateways.

Reviewer 2: @andyo-tyk This one requires MDCB 2.5.1 is it? Should we note the dependency here?

Reviewer: @caroltyk I have updated PR to mention MDCB 2.5.1. If ok and other comments ok would be to approve for release?

---

## 5.3.1 Release Notes

### Release Date TBC

### Breaking Changes
**Attention**: Please read this section carefully.

There are no breaking changes in this release, however if moving from an version of Tyk older than 5.3.0 please read the explanation provided with [5.3.0 release]({{< ref "#TykOAS-v5.3.0">}}).

### Deprecations
There are no deprecations in this release.

### Upgrade Instructions
If you are on a 5.3.0 we advise you to upgrade ASAP and if you are on an older version skip 5.3.0 and upgrade directly to this release. Go to the [Upgrading Tyk](#upgrading-tyk) section for detailed upgrade instructions.

### Release Highlights
This release primarily focuses on bug fixes.
For a comprehensive list of changes, please refer to the detailed [changelog]({{< ref "#Changelog-v5.3.1">}}) below.

### Downloads
- [Docker image to pull](TBC)
- [source code](TBC)

### Changelog {#Changelog-v5.3.1}

#### Fixed

<ul>
<li>
<details>

> <summary>Security enhancement: rejecting unsigned plugin bundles in Gateway</summary>

Author: ```suggestion
<summary>Security enhancement: don't load APIs into Gateway if custom plugin bundle fails to load</summary>
```

Issues were addressed where Tyk failed to properly reject custom plugin bundles with signature verification failures, allowing APIs to load without necessary plugins, potentially exposing upstream services. With the fix, if plugin bundle loading fails (e.g., due to signature verification), the API will not be loaded and an error will be logged in the Gateway.
</details>
</li>
<li>
<details>

> <summary>Resolved quota limit issue with API URL Rewrite to Self and Policy-Created Key</summary>

Author: ```suggestion
<summary>URL Rewrite middleware did not always correctly observe quotas for requests using keys created from policies</summary>
```

> Fixed two bugs related to quota limits not being applied and incorrect reporting of remaining quota value in the X-RateLimitRemaining header when the API had URL rewrite middleware configured with a tyk://self loop.

Author: ```suggestion
Fixed two bugs in the handling of usage quotas by the URL rewrite middleware when it was configured to rewrite to itself (e.g. to `tyk://self`). Quota limits were not observed and the quota related response headers always contained `0`. 
```

</details>
</li>
<li>
<details>

> <summary>Policy-API link deletion code updated to ensure DocumentDB compatibility</summary>

Author: ```suggestion
```

This is a Dashboard change - shouldn't be in this release note

Policy-API link deletion code was updated to address a compatibility issue with DocumentDB. The previous version relied on the $expr operator, supported by MongoDB but not DocumentDB.
</details>
</li>
<li>
<details>

> <summary>Gateway panics addressed due to JSVM usage with ignore plugin</summary>

Author: ```suggestion
<summary>Fixed a Gateway panic that could occur when using custom JavaScript plugins with the Ignore Authentication middleware</summary>
```

> Fixed a panic scenario occurring when JSVM is utilised alongside require_session:true and ignore auth middleware. While the JSVM middleware expects a valid session, the configuration flag doesn't guarantee its presence, only that it's passed if available.

Reviewer: ```suggestion
Fixed a panic scenario occurring when JSVM is utilised alongside require_session:true and ignore auth middleware. While the JSVM middleware expects a valid session, the `require_session` configuration flag doesn't guarantee its presence, only that it is passed if available.
```

Author: ```suggestion
Fixed a panic scenario that occurred when a custom JavaScript plugin that requests access to the session metadata (`require_session:true`) is assigned to the same endpoint as the Ignore Authentication middleware. While the custom plugin expects access to a valid session, the configuration flag doesn't guarantee its presence, only that it's passed if available. As such, the custom plugin should be coded to verify that the session metadata is present before attempting to use it.
```

</details>
</li>
<li>
<details>
<summary>Tyk cache confusion leads to mixed GraphQL body responses</summary>

GraphQL APIs were returning incorrect responses when simultaneous GQL calls with different inputs were made, which looked as if caching was mixed up. This was related to a setting in the GraphQL engine, that has now been turned off, so simultaneous GQL calls won't return an incorrect response.
</details>
</li>
<li>
<details>

> <summary>Gateway panics when arguments are missing in persist GraphQL endpoint operations</summary>

Reviewer: ```suggestion
<summary>Gateway panics when arguments are missing in persist GraphQL endpoints</summary>
```
